### PR TITLE
New logger rebased

### PR DESCRIPTION
**Pull request by kd0aij**
https://github.com/PX4/Firmware/pull/4045 rebased on latest master

TODO:
- [x] increased message size field to 16 bits
- [x] multi-instance topics (ongoing)
- [x] update FlightPlot and jMAVlib (ongoing)
- [ ] add parameter which lists topics to be logged with per-topic rates
- [ ] improve polling logic (eliminate usleep? perhaps a decimation factor per/topic)
- [ ] determine why SD card write blocks longer than fsync (test with nuttx_next)
- [ ] fix ROS
- [ ] SITL test
- [ ] nuttx test
- [ ] higher rate for EKF2 replay?

### CODE REVIEW

File: src/modules/logger/logger.cpp [hunk 1]
```diff
@@ -300,15 +301,24 @@ void Logger::run()
...
+					/* copy the current topic data into the buffer after the header */
 					orb_copy(sub.metadata, sub.fd, buffer + sizeof(message_data_header_s));
+
+					/* fill the message header struct in-place at the front of the buffer,
+					 * accessing the unaligned (packed) structure properly
+					 */
 					message_data_header_s *header = reinterpret_cast<message_data_header_s *>(buffer);
 					header->msg_type = static_cast<uint8_t>(MessageType::DATA);
+					/* the ORB metadata object has 2 unused trailing bytes? */
 					header->msg_size = static_cast<uint16_t>(msg_size - 2);
```
Comment (kd0aij) [inline src/modules/logger/logger.cpp]: why drop 2 bytes?

**Comment by kd0aij** [comment]
TODO:
- [ ] fix ROS
- [ ] SITL test
- [ ] nuttx test
- [x] increase message size? (currently 254)
- [ ] higher rate for EKF2 replay?

**Comment by kd0aij** [comment]
Working on logging multi-instance topics; I've added multi-subscription to the new logger, but FlightPlot is expecting only one instance to be "active" at a time. sdlog2 creates special log messages such as IMU1 for logging instance 1 of accel/gyro/mag. Should we extend FlightPlot to allow plotting of each instance of a multi-instance topic separately? That would be an improvement over creating "special" messages such as IMU1,2,3

**Comment by kd0aij** [comment]
@AndreasAntener PR for changes to ULogReader in jMAVlib: https://github.com/DrTon/jMAVlib/pull/5

**Comment by kd0aij** [comment]
First plot of a multi-instance sensor at ~100Hz. orb_set_interval seems buggy, this is using only the usleep in the logger thread for timing. Need to check uniformity of sampling interval.
accel_multi

**Comment by kd0aij** [comment]
Original conversation on new logger: https://github.com/PX4/Firmware/issues/2565#issuecomment-121958494. The format strings *are* ridiculously long.

**Comment by kd0aij** [comment]
No comments on multi-instance topics and the _unimplemented_ "active" flag?

**I propose to name all fields like this:**
```
msgFormat.name + "_N." + fieldDescr.name
```
where N is the instance ID, and eliminate the "active" field which is currently named:
```
msgFormat.name + fieldDescr.name
```

**Comment by kd0aij** [comment]
changes to the ULOG reader referenced above are in this PR: https://github.com/DrTon/jMAVlib/pull/5

**Comment by kd0aij** [comment]
@tumbili sample intervals of SENSOR_GYRO instance zero (first diff of timestamps recorded in ULOG format) x and y units are seconds:
sampleintervals

**Comment by kd0aij** [comment]
gyro_0.x resampled to a uniform 100Hz in Octave:

resampled

**Comment by kd0aij** [comment]
First use of new logger to calculate angular rate impulse response (from rate setpoint to gyro rate). Quadcopter is a Spedix S250AQ with PixRacer:

rateresponse

**Comment by tumbili** [comment]
@kd0aij Cool! Is there a chance to plot this logarithmic on both axes (bode plot)?

**Comment by tumbili** [comment]
@kd0aij Also it would be nice to have the same data for the actual plant, meaning transfer function form roll / pitch control to roll rate.

**Comment by kd0aij** [comment]
@tumbili Working on the RC input -> rate response next. loglog plot:

rateresponse

**Comment by kd0aij** [comment]
first shot at acro mode RC input to rate response; looks plausible

rc_acroresponse

**Comment by kd0aij** [comment]
ACRO mode test flight with roll/pitch P gains cranked up to 0.2. Orientation hold was noticeably better, and there was no visible or audible oscillation:
.ulog and CSV files: https://drive.google.com/open?id=0Bw3digSMQXDuZmpRZlBmY0p6bDA

impresponse

**Comment by kd0aij** [comment]
Plot of rollrate tracking with the high P gain from log above. Latency is only 20msec from setpoint to gyro rate:

rollratetracking

**Comment by LorenzMeier** [comment]
Nice! Could you reconnect with the testers on http://discuss.px4.io and see if they like this better?

**Comment by tumbili** [comment]
@kd0aij So we same out with the right conclusion of our analysis? :-D One more thing, could you maybe not scale the magnitude plot such that the maximum is always 1 (0dB)?
Also what would be nice (if you have time) is the open loop transfer function (angle error to output).

**Comment by tumbili** [comment]
@kd0aij Could you also specify what you mean by the 20ms latency?

**Comment by tumbili** [comment]
@LorenzMeier After the conference I would like to help @kd0aij finish this off.

**Comment by kd0aij** [comment]
@tumbili I updated the last transfer function plot above. Getting close to a "proper" Bode plot now, I think.

**Comment by tumbili** [comment]
@kd0aij Thanks that looks good.

**Comment by tumbili** [comment]
@kd0aij Any change to get the open loop transfter function? I guess you just have to replace the rate setpoint with the rate error.

**Comment by kd0aij** [comment]
what I mean by 20msec latency is the delay from control input to gyro rates. I'm calculating it by cross-correlating the rates with the control signal: (1 unit of lag is one sample interval = 8msec)

xcorr

**Comment by kd0aij** [comment]
For open-loop response, I'll try using rateError := (rate - rateSP) as the input

**Comment by tumbili** [comment]
@kd0aij Cool, that would be great!

**Comment by kd0aij** [comment]
roll and pitch responses matched up with PID tuning:

rate_impulse

**Comment by kd0aij** [comment]
first measurement of attitude control latency (setpoint to IMU attitude); red line is setpoint compensated for measured latency of about 100msec, green line is IMU attitude:

att_latency

**Comment by kd0aij** [comment]
superseded by https://github.com/PX4/Firmware/pull/4197

**Pull request closed without merging**